### PR TITLE
On Click BlockData

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR aims to support blockdata for on click

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #6305

### CODE REVIEW

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 1]
```diff
@@ -62,30 +45,29 @@ public class EvtClick extends SkriptEvent {
...
 						"Please note that rightclick events with an empty hand while not looking at a block are not sent to the server, so there's no way to detect them.",
 						"Also note that a leftclick on an entity is an attack and thus not covered by the 'click' event, but the 'damage' event.")
 				.examples("on click:",
 						"on rightclick holding a fishing rod:",
 						"on leftclick on a stone or obsidian:",
 						"on rightclick on a creeper:",
-						"on click with a sword:")
-				.since("1.0");
+						"on click with a sword:",
+						"on click on chest[facing=north]:")
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/events/EvtClick.java:58]:
```suggestion
						"on click on chest[facing=north]:",
						"on click on campfire[lit=true]")
```

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 2]
```diff
@@ -62,30 +45,29 @@ public class EvtClick extends SkriptEvent {
...
 
 	/**
 	 * Only trigger when one of these is interacted with.
 	 */
-	@Nullable
-	private Literal<?> type;
+	private @Nullable Literal<?> type;
 
 	/**
 	 * Only trigger when then item player clicks with is one of these.
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/events/EvtClick.java:68]:
```suggestion
	 * Only trigger when the item that the player clicks with is one of these.
```

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 3]
```diff
@@ -187,8 +166,8 @@ public boolean check(Event event) {
 		if (tools != null && !tools.check(event, new Checker<ItemType>() {
 			@Override
 			public boolean check(final ItemType t) {
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/events/EvtClick.java:168]:
```suggestion
			public boolean check(final ItemType itemType) {
```

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 4]
```diff
@@ -142,18 +123,16 @@ public boolean check(Event event) {
...
 			
 			entity = clicked;
 			block = null;
-		} else if (event instanceof PlayerInteractEvent) {
-			PlayerInteractEvent clickEvent = ((PlayerInteractEvent) event);
-			
+		} else if (event instanceof PlayerInteractEvent interactEvent) {
 			// Figure out click type, filter non-click events
-			Action a = clickEvent.getAction();
+			Action a = interactEvent.getAction();
```
Comment (cheeezburga) [inline src/main/java/ch/njol/skript/events/EvtClick.java:135]:
```suggestion
			Action action = interactEvent.getAction();
```

**Review by cheeezburga — CHANGES_REQUESTED** [review]
Just some small stuff, but you should also remove `final`s that are scattered throughout, and theres still code that can be updated (switch statements, etc.).

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 5]
```diff
@@ -201,23 +180,27 @@ public boolean check(final ItemType t) {
 		}
 		
 		if (type != null) {
+			BlockData blockDataCheck = block != null ? block.getBlockData() : null;
 			return type.check(event, new Checker<Object>() {
 				@Override
 				public boolean check(final Object o) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtClick.java:186]:
```suggestion
				public boolean check(Object object) {
```

**Review by Efnilite — APPROVED** [review]
looks good with @cheeezburga's changes

**Comment by TheAbsolutionism** [comment]
> Just some small stuff, but you should also remove `final`s that are scattered throughout, and theres still code that can be updated (switch statements, etc.).

I could change to an enhanced switch, but I cant put both `case PHYSICAL` and `default` on the same line leading to two `return false`s
Other than that, I dont really see anything else sticking out that might need to be updated, imo, but if you would like to point out some things I can take a look.

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 6]
```diff
@@ -50,42 +33,42 @@ public class EvtClick extends SkriptEvent {
...
 	 * Click types.
 	 */
-	private final static int RIGHT = 1, LEFT = 2, ANY = RIGHT | LEFT;
+	private static int RIGHT = 1, LEFT = 2, ANY = RIGHT | LEFT;
 
 	/**
 	 * Tracks PlayerInteractEvents to deduplicate them.
 	 */
-	public static final ClickEventTracker interactTracker = new ClickEventTracker(Skript.getInstance());
+	public static ClickEventTracker interactTracker = new ClickEventTracker(Skript.getInstance());
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/events/EvtClick.java:41]: The final can remain on fields. This is not related to the common final abuse of parameters seen in Skript coding standards.
```suggestion
	public final static ClickEventTracker interactTracker = new ClickEventTracker(Skript.getInstance());
```

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 7]
```diff
@@ -50,42 +33,42 @@ public class EvtClick extends SkriptEvent {
 	/**
 	 * Click types.
 	 */
-	private final static int RIGHT = 1, LEFT = 2, ANY = RIGHT | LEFT;
+	private static int RIGHT = 1, LEFT = 2, ANY = RIGHT | LEFT;
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/events/EvtClick.java:36]: The final can remain on fields. This is not related to the common final abuse of parameters seen in Skript coding standards.
```suggestion
	private final static int RIGHT = 1, LEFT = 2, ANY = RIGHT | LEFT;
```

**Review by TheLimeGlass — APPROVED** [review]

**Review by cheeezburga — APPROVED** [review]

**Comment by cheeezburga** [comment]
> I could change to an enhanced switch, but I cant put both `case PHYSICAL` and `default` on the same line leading to two `return false`s Other than that, I dont really see anything else sticking out that might need to be updated, imo, but if you would like to point out some things I can take a look.

I think an enhanced one would look a bit nicer, but up to you. I don't think there was anything else, but I wouldn't hate if you wanted to change the error messages a bit in `init`, as they seem a bit clunky and/or misleading at the moment.

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 8]
```diff
@@ -97,22 +80,22 @@ public class EvtClick extends SkriptEvent {
...
-					Skript.error("A leftclick on an entity is an attack and thus not covered by the 'click' event, but the 'vehicle damage' event. " +
-							"Change this event to a rightclick to fix this warning message.");
+					Skript.error("A leftclick on a vehicle entity is an attack and thus not covered by the 'click' event, use the 'vehicle damage' event for left clicks " +
+							"or change this event to a rightclick.");
 				} else {
-					Skript.error("A leftclick on an entity is an attack and thus not covered by the 'click' event, but the 'damage' event. " +
-							"Change this event to a rightclick to fix this warning message.");
+					Skript.error("A leftclick on an entity is an attack and thus not covered by the 'click' event, use the 'damage' event for left clicks " +
+							"or change this event to a rightclick.");
 				}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:99]: These were fine; I don't think they need changes beyond the addition of `a vehicle` like the change above.

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 9]
```diff
@@ -186,38 +163,42 @@ public boolean check(Event event) {
...
+				if (event instanceof PlayerInteractEvent interactEvent) {
+					return itemType.isOfType(interactEvent.getItem());
 				} else { // PlayerInteractEntityEvent doesn't have item associated with it
 					PlayerInventory invi = ((PlayerInteractEntityEvent) event).getPlayer().getInventory();
 					ItemStack item = ((PlayerInteractEntityEvent) event).getHand() == EquipmentSlot.HAND
 							? invi.getItemInMainHand() : invi.getItemInOffHand();
-					return t.isOfType(item);
+					return itemType.isOfType(item);
 				}
 			}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:175]: the checker could just be a lambda

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 10]
```diff
@@ -97,22 +80,22 @@ public class EvtClick extends SkriptEvent {
...
 			} else if (click == ANY) {
 				if (Vehicle.class.isAssignableFrom(entitydata.getSingle().getType())) {
-					Skript.error("A leftclick on an entity is an attack and thus not covered by the 'click' event, but the 'vehicle damage' event. " +
-							"Change this event to a rightclick to fix this warning message.");
+					Skript.error("A leftclick on a vehicle entity is an attack and thus not covered by the 'click' event, but the 'vehicle damage' event. " +
+						"Change this event to a rightclick to fix this warning message.");
 				} else {
 					Skript.error("A leftclick on an entity is an attack and thus not covered by the 'click' event, but the 'damage' event. " +
-							"Change this event to a rightclick to fix this warning message.");
+						"Change this event to a rightclick to fix this warning message.");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:98]: why were the tabs removed? a continuation of a line should have two tabs following it re: coding conventions

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 11]
```diff
@@ -142,82 +124,76 @@ public boolean check(Event event) {
...
-					return t.isOfType(item);
-				}
+
+		if (tools != null && !tools.check(event, itemType -> {
+			if (event instanceof PlayerInteractEvent interactEvent) {
+				return itemType.isOfType(interactEvent.getItem());
+			} else {
+				PlayerInventory invi = ((PlayerInteractEntityEvent) event).getPlayer().getInventory();
+				ItemStack item = ((PlayerInteractEntityEvent) event).getHand() == EquipmentSlot.HAND
+					? invi.getItemInMainHand() : invi.getItemInOffHand();
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:170]: same here, two tabs

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 12]
```diff
@@ -142,82 +124,76 @@ public boolean check(Event event) {
...
+			BlockData blockDataCheck = block != null ? block.getBlockData() : null;
 			return type.check(event, new Checker<Object>() {
 				@Override
-				public boolean check(final Object o) {
+				public boolean check(Object object) {
 					if (entity != null) {
-						return o instanceof EntityData ? ((EntityData<?>) o).isInstance(entity) : Relation.EQUAL.isImpliedBy(DefaultComparators.entityItemComparator.compare(EntityData.fromEntity(entity), (ItemType) o));
-					} else {
-						return o instanceof EntityData ? false : ((ItemType) o).isOfType(block);
+						return object instanceof EntityData ? ((EntityData<?>) object).isInstance(entity) : Relation.EQUAL.isImpliedBy(DefaultComparators.entityItemComparator.compare(EntityData.fromEntity(entity), (ItemType) object));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:181]: should split over multiple lines, very far over the 180 char limit

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 13]
```diff
@@ -142,82 +124,76 @@ public boolean check(Event event) {
...
 			});
 		}
 		return true;
 	}
 
 	@Override
-	public String toString(@Nullable Event e, boolean debug) {
-		return (click == LEFT ? "left" : click == RIGHT ? "right" : "") + "click" + (type != null ? " on " + type.toString(e, debug) : "") + (tools != null ? " holding " + tools.toString(e, debug) : "");
+	public String toString(@Nullable Event event, boolean debug) {
+		return (click == LEFT ? "left" : click == RIGHT ? "right" : "") + "click" + (type != null ? " on " + type.toString(event, debug) : "") + (tools != null ? " holding " + tools.toString(event, debug) : "");
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:196]: also split over multiple lines

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 14]
```diff
@@ -142,82 +124,86 @@ public boolean check(Event event) {
...
-				default:
+			switch (action)  {
+				case LEFT_CLICK_AIR, LEFT_CLICK_BLOCK -> click = LEFT;
+				case RIGHT_CLICK_AIR, RIGHT_CLICK_BLOCK -> click = RIGHT;
+				case PHYSICAL -> {
+					return false;
+				}
+				default -> {
 					return false;
+				}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:146]: can't these be collapsed into one?

File: src/main/java/ch/njol/skript/events/EvtClick.java [hunk 15]
```diff
@@ -142,82 +124,86 @@ public boolean check(Event event) {
...
+				PlayerInventory invi = ((PlayerInteractEntityEvent) event).getPlayer().getInventory();
+				ItemStack item = ((PlayerInteractEntityEvent) event).getHand() == EquipmentSlot.HAND
+						? invi.getItemInMainHand() : invi.getItemInOffHand();
+				return itemType.isOfType(item);
 			}
-		})) {
-			return false;
-		}
-		
+		})) { return false; }
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/events/EvtClick.java:173]:
```suggestion
		})) {
			return false;
		}
```

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/events/EvtClick.java:173]: actually you might want to declare the lambda as a predicate prior to the if statement, so the if statement doesn't get lost in the sauce.

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Pull request merged**